File: examples/pi_remote.py
"""
pi-remote Python client — minimal SSE parser. Copy-paste into your project.

    client = PiRemote("http://localhost:8080")
    result = client.chat("fix the bug in auth.ts")
    print(result["text"])

No pip install beyond `requests`.
"""

import sys, json

try:
    import requests
except ImportError:
    print("pip install requests", file=sys.stderr); sys.exit(1)

class PiRemote:
    def __init__(self, url="http://localhost:8080", api_key=None):
        self.url = url.rstrip("/")
        self.api_key = api_key

    def _headers(self):
        h = {"Content-Type": "application/json"}
        if self.api_key: h["Authorization"] = f"Bearer {self.api_key}"
        return h

    def health(self):
        return requests.get(f"{self.url}/v1/health", headers=self._headers()).json()

    def session(self):
        return requests.post(f"{self.url}/v1/sessions", json={}, headers=self._headers()).json()

    def chat(self, message, session_id=None, on_token=None, on_tool=None):
        """Send a prompt, stream SSE, return dict with text + tool_calls."""
        body = {"message": message}
        if session_id: body["sessionId"] = session_id

        r = requests.post(f"{self.url}/v1/chat", json=body, headers=self._headers(), stream=True)

        result = {"text": "", "tool_calls": []}
        current_event = None

        for line in r.iter_lines(decode_unicode=True):
            if not line: continue
            if line.startswith("event: "):
                current_event = line[7:]
            elif line.startswith("data: "):
                data = json.loads(line[6:])
                if current_event == "token":
                    t = data.get("text", ""); result["text"] += t
                    sys.stdout.write(t); sys.stdout.flush()
                    if on_token: on_token(t)
                elif current_event == "tool_start":
                    call = {"name": data.get("tool"), "args": data.get("args")}
                    result["tool_calls"].append(call)
                    if on_tool: on_tool(call)
                elif current_event == "done":
                    result["session_id"] = data.get("sessionId")
                    break

        sys.stdout.write("\n"); sys.stdout.flush()
        return result


if __name__ == "__main__":
    import argparse
    p = argparse.ArgumentParser(description="pi-remote chat")
    p.add_argument("prompt", nargs="?", default="say hello")
    p.add_argument("--url", default="http://localhost:8080")
    p.add_argument("--key", default=None)
    args = p.parse_args()
    client = PiRemote(args.url, args.key)
    print(f"pi-remote {args.url} — {client.health()['status']}\n")
    client.chat(args.prompt)

File: examples/pi_remote.mjs
#!/usr/bin/env node
/**
 * pi-remote JavaScript client — minimal SSE parser. Copy-paste into your project.
 *
 *   const client = new PiRemote("http://localhost:8080");
 *   const result = await client.chat("fix the bug");
 *   console.log(result.text);
 *
 * No npm install. Node 18+.
 */

class PiRemote {
  constructor(url = "http://localhost:8080", apiKey = null) {
    this.url = url.replace(/\/$/, "");
    this.apiKey = apiKey;
  }

  #headers() {
    const h = { "Content-Type": "application/json" };
    if (this.apiKey) h["Authorization"] = `Bearer ${this.apiKey}`;
    return h;
  }

  async health() {
    const r = await fetch(`${this.url}/v1/health`, { headers: this.#headers() });
    return r.json();
  }

  async session() {
    const r = await fetch(`${this.url}/v1/sessions`, { method: "POST", headers: this.#headers() });
    return r.json();
  }

  async chat(message, { sessionId, onToken, onTool } = {}) {
    const body = sessionId ? { message, sessionId } : { message };
    const res = await fetch(`${this.url}/v1/chat`, {
      method: "POST", headers: this.#headers(), body: JSON.stringify(body),
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${await res.text()}`);

    const result = { text: "", toolCalls: [] };
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "", currentEvent = "", finished = false;

    while (!finished) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        if (line.startsWith("event: ")) {
          currentEvent = line.slice(7);
        } else if (line.startsWith("data: ") && currentEvent === "token") {
          const data = JSON.parse(line.slice(6));
          const t = data.text ?? "";
          result.text += t;
          process.stdout.write(t);
          onToken?.(t);
        } else if (line.startsWith("data: ") && currentEvent === "tool_start") {
          const data = JSON.parse(line.slice(6));
          result.toolCalls.push({ name: data.tool, args: data.args });
          onTool?.({ name: data.tool, args: data.args });
        } else if (line.startsWith("data: ") && currentEvent === "done") {
          finished = true;
          break;
        }
      }
    }

    process.stdout.write("\n");
    return result;
  }
}

// -- demo --
const args = process.argv.slice(2);
let url = "http://localhost:8080";
let prompt = "say hello";
for (let i = 0; i < args.length; i++) {
  if (args[i] === "--url" && args[i + 1]) url = args[++i];
  else if (args[i] === "--key" && args[i + 1]) args[++i]; // skip key
  else { prompt = args.slice(i).join(" "); break; }
}

const client = new PiRemote(url);
console.log(`pi-remote ${url} — ${(await client.health()).status}\n`);
await client.chat(prompt);
